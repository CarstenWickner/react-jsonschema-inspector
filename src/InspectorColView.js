import PropTypes from "prop-types";
import React, { PureComponent } from "react";

import InspectorColumn from "./InspectorColumn";
import JsonSchemaPropType from "./JsonSchemaPropType";

class InspectorColView extends PureComponent {
    componentDidUpdate(prevProps) {
        const previousColumnCount = prevProps.columnData.length;
        // eslint-disable-next-line react/destructuring-assignment
        const currentColumnCount = this.props.columnData.length;
        if (previousColumnCount < currentColumnCount) {
            // auto-scroll to the far right if an additional column was added
            this.colViewContainerRef.scrollLeft = this.colViewContainerRef.scrollWidth;
        }
    }

    render() {
        const {
            columnData, refTargets, appendEmptyColumn, renderItemContent
        } = this.props;
        return (
            <div className="jsonschema-inspector-colview" ref={(ref) => { this.colViewContainerRef = ref; }}>
                {columnData.map((singleColumnData, index) => (
                    <InspectorColumn
                        // eslint-disable-next-line react/no-array-index-key
                        key={index}
                        refTargets={refTargets}
                        renderItemContent={renderItemContent}
                        {...singleColumnData}
                    />
                ))}
                {appendEmptyColumn
                    && <div className="jsonschema-inspector-column-placeholder" />}
            </div>
        );
    }
}

InspectorColView.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool
    })).isRequired,
    refTargets: PropTypes.objectOf(JsonSchemaPropType).isRequired,
    appendEmptyColumn: PropTypes.bool,
    renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema, refTargets })
};

InspectorColView.defaultProps = {
    appendEmptyColumn: false,
    renderItemContent: null
};

export default InspectorColView;
