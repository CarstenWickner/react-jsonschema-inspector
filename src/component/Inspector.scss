.jsonschema-inspector {
    display: flex;
    flex-direction: column;
    font-size: 13px;
    font-family: Arial, Helvetica, sans-serif;

    $columnWidth: 15em;
    $detailsWidth: 20em;
    $selectionColor: #dcdcdc;
    $trailingSelectionColor: #276bd2;
    $backgroundColor: #fff;
    $innerBorderColor: #e0e0e0;
    $outerBorderColor: #cbcbcb;
    $innerBorder: 1px solid $innerBorderColor;
    $outerBorder: 1px solid $outerBorderColor;

    $verticalSpacingOutside: 0.25em;
    $verticalItemSpacingInside: 0.25em;
    $leadItemSpacingInside: 0.75em;
    $spacingInside: 0.5em;
    $borderRadius: 0.25em;

    &-header,
    &-body,
    &-footer {
        display: flex;
        background-color: $backgroundColor;
        border: $outerBorder;
        border-radius: $borderRadius;
    }
    &-body {
        height: 25em;
    }
    &-colview {
        flex-grow: 1;
        display: flex;
        overflow-x: auto;
        margin-right: -1px;
    }
    &-column {
        &,
        &-placeholder {
            display: flex;
            flex-direction: column;
            width: $columnWidth;
            min-width: $columnWidth;
            border-right: $innerBorder;
            overflow-y: auto;
        }
        &-placeholder {
            opacity: 0;
        }
        .optional-group-title {
            display: block;
            padding: $verticalItemSpacingInside $spacingInside $verticalItemSpacingInside $leadItemSpacingInside;
            font-style: italic;
        }
        &.optional-groups ul {
            margin: 0;
            padding: 0;
            list-style-type: none;

            ul {
                background-color: $innerBorderColor;
            }
            ul,
            .jsonschema-inspector-item {
                padding-left: $leadItemSpacingInside;
            }
            ul .optional-group-title,
            ul .jsonschema-inspector-item {
                background-color: $backgroundColor;
            }
            .optional-group-title {
                padding-left: (2 * $leadItemSpacingInside);
            }
        }
    }
    &-item {
        flex-shrink: 0;
        text-align: start;
        border: none;
        background: inherit;
        color: inherit;
        font: inherit;
        padding: 0;
        margin: 0;
        width: 100%;

        &.selected {
            background-color: $selectionColor;
        }
        &.matching-filter {
            font-weight: bold;
        }
        &.not-matching-filter {
            opacity: 0.5;
        }
        &-content {
            display: flex;
            flex-direction: row;
        }
        &-name {
            display: block;
            flex-grow: 1;
            padding: $verticalItemSpacingInside $spacingInside $verticalItemSpacingInside $leadItemSpacingInside;
        }
        &-icon {
            display: block;
            flex-grow: 0;
            width: 1em;
        }
        &.has-nested-items > &-content > &-icon {
            background-color: #7f7f7f;
            -webkit-clip-path: polygon(25% 25%, 75% 50%, 25% 75%);
            clip-path: polygon(25% 25%, 75% 50%, 25% 75%);
        }
    }
    &-column.trailing-selection &-item.selected {
        background-color: $trailingSelectionColor;
        color: white;
    }
    &-column.trailing-selection &-item.selected &-item-icon:after {
        color: inherit;
    }
    &-details {
        width: $detailsWidth;
        min-width: $detailsWidth;
        overflow: auto;
        border-left: $innerBorder;

        &-header:first-child {
            margin-top: 0;
        }
        &-content {
            padding: 1em;
        }
        &-form {
            display: flex;
            flex-direction: column;
            $labelWidth: 30%;

            &-row {
                display: flex;
                flex-direction: row;
                margin-bottom: 0.5em;
            }
            &-label,
            &-value {
                display: inline-block;
                vertical-align: top;
                word-wrap: break-word;
            }
            &-label {
                width: $labelWidth;
                padding-right: 0.5em;
            }
            &-value {
                width: 100% - $labelWidth;
            }
        }
    }
    &-header {
        flex-direction: row-reverse;
        margin-bottom: $verticalSpacingOutside;
    }
    &-search {
        padding: $spacingInside;

        &-input {
            font: inherit;
            line-height: 2em;   
            border: $innerBorder;
            border-radius: $borderRadius;
            padding: 0 $spacingInside;
            min-width: ($detailsWidth - (2 * $spacingInside));
        }
    }
    &-footer {
        margin-top: $verticalSpacingOutside;
    }
    &-breadcrumbs {
        display: flex;
        flex-flow: row wrap;
        padding: $spacingInside 0.75em $spacingInside (5 * $spacingInside);
        font-family: 'Courier New', Courier, monospace;

        &-icon,
        &-item {
            display: block;
            border-radius: $borderRadius;
        }
        &-icon {
            background-color: $trailingSelectionColor;
            height: 1.5em;
            width: 0.8em;
            margin: 0.25em 0 0.25em (-4 * $spacingInside);
            -webkit-clip-path: polygon(0% 0%, 100% 50%, 0% 100%);
            clip-path: polygon(0% 0%, 100% 50%, 0% 100%);
        }
        &-item {
            background-color: $selectionColor;
            margin-right: -0.75em;
            padding: 0 0.25em 0 1.25em;
            line-height: 2em;
            text-overflow: ellipsis;

            &.has-nested-items {
                padding-right: 1em;
                -webkit-clip-path: polygon(0% 5%, calc(100% - 1em) 5%, 100% 50%, calc(100% - 1em) 95%, 0% 95%, 1em 50%);
                clip-path: polygon(0% 5%, calc(100% - 1em) 5%, 100% 50%, calc(100% - 1em) 95%, 0% 95%, 1em 50%);
            }
            &:not(.has-nested-items) {
                -webkit-clip-path: polygon(0% 5%, 100% 5%, 100% 95%, 0% 95%, 1em 50%);
                clip-path: polygon(0% 5%, 100% 5%, 100% 95%, 0% 95%, 1em 50%);
            }
        }
    }
}
