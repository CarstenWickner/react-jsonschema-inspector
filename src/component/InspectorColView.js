import PropTypes from "prop-types";
import React, { Component } from "react";

import InspectorColumn from "./InspectorColumn";
import InspectorOptionsColumn from "./InspectorOptionsColumn";
import { getColumnDataPropTypeShape } from "./renderDataUtils";

class InspectorColView extends Component {
    componentDidUpdate(prevProps) {
        const previousColumnCount = prevProps.columnData.length + (prevProps.appendEmptyColumn ? 1 : 0);
        const { columnData, appendEmptyColumn } = this.props;
        const currentColumnCount = columnData.length + (appendEmptyColumn ? 1 : 0);
        if (previousColumnCount < currentColumnCount) {
            // auto-scroll to the far right if an additional column was added
            this.colViewContainerRef.scrollLeft = this.colViewContainerRef.scrollWidth;
        }
    }

    render() {
        const {
            columnData, appendEmptyColumn, renderItemContent
        } = this.props;
        return (
            <div
                className="jsonschema-inspector-colview"
                ref={(ref) => { this.colViewContainerRef = ref; }}
                tabIndex={-1}
            >
                {columnData.map((singleColumnData, index) => {
                    const {
                        items, options, contextGroup, selectedItem, trailingSelection, filteredItems, onSelect
                    } = singleColumnData;
                    if (items) {
                        return (
                            <InspectorColumn
                                // eslint-disable-next-line react/no-array-index-key
                                key={index}
                                items={items}
                                selectedItem={selectedItem}
                                trailingSelection={trailingSelection}
                                filteredItems={filteredItems}
                                onSelect={onSelect}
                                renderItemContent={renderItemContent}
                            />
                        );
                    }
                    return (
                        <InspectorOptionsColumn
                            // eslint-disable-next-line react/no-array-index-key
                            key={index}
                            options={options}
                            contextGroup={contextGroup}
                            selectedItem={selectedItem}
                            trailingSelection={trailingSelection}
                            filteredItems={filteredItems}
                            onSelect={onSelect}
                            renderItemContent={renderItemContent}
                        />
                    );
                })}
                {appendEmptyColumn
                    && <div className="jsonschema-inspector-column-placeholder" />}
            </div>
        );
    }
}

InspectorColView.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(true))).isRequired,
    appendEmptyColumn: PropTypes.bool,
    renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
};

InspectorColView.defaultProps = {
    appendEmptyColumn: false,
    renderItemContent: null
};

export default InspectorColView;
