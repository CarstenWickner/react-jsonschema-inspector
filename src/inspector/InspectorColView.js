import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import InspectorColumn from './InspectorColumn';
import JsonSchemaPropType from './JsonSchemaPropType';

class InspectorColView extends PureComponent {
    render() {
        const { columnData, renderItemContent } = this.props;
        return (
            <div className="jsonschema-inspector-colview">
                {columnData.map((singleColumnData, index) => (
                    <InspectorColumn
                        key={index}
                        renderItemContent={renderItemContent}
                        {...singleColumnData}
                    />
                ))}
            </div>
        );
    }
}

InspectorColView.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool,
        refTargets: PropTypes.objectOf(JsonSchemaPropType)
    })),
    renderItemContent: PropTypes.func // func(string: name, JsonSchema: schema, boolean: selected)
}

export default InspectorColView;