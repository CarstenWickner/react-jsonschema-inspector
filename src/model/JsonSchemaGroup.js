import { listValues } from "./utils";

/**
 * Representation of an array of schemas (e.g. `allOf`, `anyOf`, `oneOf`), offering a number of convenience functions for extracting information.
 */
export default class JsonSchemaGroup {
    /**
     * @type {Array.<JsonSchema|JsonSchemaGroup>}
     */
    entries = [];

    /**
     * Determine whether an entry of type `JsonSchema` should be treated as a separate selectable option (i.e. like a `JsonSchemaGroup`).
     *
     * @returns {boolean} by default: always 'false', but may be overridden by sub-classes
     */
    // eslint-disable-next-line class-methods-use-this
    considerSchemasAsSeparateOptions() {
        return false;
    }

    /**
     * Indicate whether the entries of this group should be treated as if they were all defined in a single schema.
     * Otherwise, all entries shall be listed as alternative options to choose from.
     *
     * @returns {boolean} whether entries should be included transparently (otherwise as options)
     */
    shouldTreatEntriesAsOne() {
        let filteredEntries = this.entries;
        if (!this.considerSchemasAsSeparateOptions()) {
            filteredEntries = filteredEntries.filter((entry) => entry instanceof JsonSchemaGroup && !entry.shouldTreatEntriesAsOne());
        }
        return filteredEntries.length < 2;
    }

    /**
     * Adding the given Json Schema or group to this group.
     *
     * @param {JsonSchema|JsonSchemaGroup} schemaOrGroup - entry to add to this group
     * @returns {JsonSchemaGroup} this (i.e. self-reference for chaining)
     */
    with(schemaOrGroup) {
        if (schemaOrGroup instanceof JsonSchemaGroup && schemaOrGroup.entries.length === 1) {
            // unwrap a group containing only a single entry
            this.entries.push(schemaOrGroup.entries[0]);
        } else {
            this.entries.push(schemaOrGroup);
        }
        return this;
    }

    /**
     * Iterate through the parts of this schema group until the given predicate `checkEntry` returns 'true' for one.
     *
     * @param {Function} checkEntry - predicate to invoke for each schema part until 'true' is returned
     * @param {JsonSchema} checkEntry.param0 - single plain schema to check (potentially ignoring any nested `allOf`/`oneOf`/`anyOf`)
     * @param {boolean} checkEntry.param1 - whether nested `allOf`/`oneOf`/`anyOf` may be included in the check
     * @param {boolean} checkEntry.return - whether the predicate's condition was fulfilled
     * @param {?Array.<{index: number}>} optionTarget - array of mutable objects indicating which optional schema parts to consider (ignoring others)
     * @param {number} optionTarget[].index - mutable index, a value of `0` marks the optional schema part to be considered
     * @returns {boolean} whether `checkEntry` returned 'true' for any item in this group's `entries`
     */
    someEntry(checkEntry, optionTarget) {
        const considerSchemasAsOptions = this.considerSchemasAsSeparateOptions();
        const treatEntriesAsSingleSchema = this.shouldTreatEntriesAsOne();
        return this.entries.some((entry) => {
            if (!treatEntriesAsSingleSchema && (considerSchemasAsOptions || entry instanceof JsonSchemaGroup)) {
                // an optional entry should be skipped if it is not the specifically selected one
                const isSelectedEntry = !optionTarget || (optionTarget.length && optionTarget[0].index === 0);
                if (optionTarget && optionTarget.length) {
                    // eslint-disable-next-line no-param-reassign
                    optionTarget[0].index -= 1;
                }
                if (!isSelectedEntry) {
                    // ignore unselected option, i.e. indicate continued traversing/not to stop
                    return false;
                }
            }
            if (entry instanceof JsonSchemaGroup) {
                // recursively check all parts of this nested group
                return entry.someEntry(
                    checkEntry,
                    (treatEntriesAsSingleSchema || !optionTarget) ? optionTarget : optionTarget.slice(1)
                );
            }
            // entry is a JsonSchema, invoke the given checkEntry function and return its result
            return checkEntry(entry, !optionTarget || !optionTarget.length);
        });
    }

    /**
     * Extract the properties mentioned in this schema group.
     *
     * @param {Function} extractFromSchema - mapping function to invoke for extracting value(s) from a single JsonSchema
     * @param {JsonSchema} extractFromSchema.param0 - single JsonSchema to extract value(s) from
     * @param {*} extractFromSchema.return - extracted value(s) from a single JsonSchema
     * @param {?Function} mergeResults - function to be used in `Array.reduce()` to combine extracted values from multiple JsonSchemas
     * @param {*} mergeResults.param0 - combined values
     * @param {*} mergeResults.param1 - single value to add to the already combined values
     * @param {*} mergeResults.return - combined values including additional single value
     * @param {*} defaultValue - initial value of mergeResults.param0 on first execution
     * @param {?Array.<{index: number}>} optionTarget - array of mutable objects indicating which optional schema parts to consider (ignoring others)
     * @param {number} optionTarget[].index - mutable index, a value of `0` marks the optional schema part to be considered
     * @returns {*} return combined extracted values from this schema group
     */
    extractValues(extractFromSchema, mergeResults = listValues, defaultValue, optionTarget) {
        // collect schemas where we have both a boolean and a proper schema, favour the proper schema
        const values = [];
        const addToResultAndContinue = (entry) => {
            values.push(extractFromSchema(entry));
            // indicate to continue traversing all entries
            return false;
        };
        this.someEntry(addToResultAndContinue, optionTarget);
        return values.reduce(mergeResults, defaultValue);
    }

    /**
     * Create representation of this group's given options.
     *
     * @param {Array.<{groupTitle: ?string, options: ?Array.<object>, nameForIndex: ?Function}>} containedOptions - nested option representations
     * @returns {{groupTitle: ?string, options: ?Array.<object>, nameForIndex: ?Function}} representation of the given group's top level options
     */
    // eslint-disable-next-line class-methods-use-this
    createOptionsRepresentation(containedOptions) {
        let result;
        if (containedOptions.length === 0) {
            result = {};
        } else if (containedOptions.length === 1) {
            // remove unnecessary hierarchy level by simply returning the single option from the array directly
            [result] = containedOptions;
        } else {
            result = {
                options: containedOptions
            };
        }
        return result;
    }
}
