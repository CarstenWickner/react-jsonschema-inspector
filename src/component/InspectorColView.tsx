import PropTypes from "prop-types";
import React, { Component } from "react";

import InspectorColumn from "./InspectorColumn";
import InspectorOptionsColumn from "./InspectorOptionsColumn";
import { getColumnDataPropTypeShape } from "./renderDataUtils";
import { RenderItemsColumn, RenderOptionsColumn, RenderItemContentFunction } from "../types/Inspector";

interface ColViewDefaultProps {
    appendEmptyColumn: boolean,
    renderItemContent: RenderItemContentFunction
};

interface ColViewProps extends ColViewDefaultProps {
    columnData: Array<RenderItemsColumn | RenderOptionsColumn>
};

class InspectorColView extends Component<ColViewProps> {
    private colViewContainerRef: React.RefObject<HTMLDivElement>;

    constructor(props: ColViewProps) {
        super(props);
        this.colViewContainerRef = React.createRef();
    }

    componentDidUpdate(prevProps: ColViewProps) {
        const previousColumnCount = prevProps.columnData.length + (prevProps.appendEmptyColumn ? 1 : 0);
        const { columnData, appendEmptyColumn } = this.props;
        const currentColumnCount = columnData.length + (appendEmptyColumn ? 1 : 0);
        if (previousColumnCount < currentColumnCount) {
            // auto-scroll to the far right if an additional column was added
            this.colViewContainerRef.current.scrollLeft = this.colViewContainerRef.current.scrollWidth;
        }
    }

    render() {
        const {
            columnData, appendEmptyColumn, renderItemContent
        } = this.props;
        return (
            <div
                className="jsonschema-inspector-colview"
                ref={this.colViewContainerRef}
                tabIndex={-1}
            >
                {columnData.map((singleColumnData, index) => {
                    const {
                        selectedItem, trailingSelection, filteredItems, onSelect
                    } = singleColumnData;
                    if ((singleColumnData as RenderItemsColumn).items) {
                        return (
                            <InspectorColumn
                                // eslint-disable-next-line react/no-array-index-key
                                key={index}
                                items={(singleColumnData as RenderItemsColumn).items}
                                selectedItem={selectedItem as string}
                                trailingSelection={trailingSelection}
                                filteredItems={filteredItems as Array<string>}
                                onSelect={onSelect}
                                renderItemContent={renderItemContent}
                            />
                        );
                    }
                    const { options, contextGroup } = singleColumnData as RenderOptionsColumn;
                    return (
                        <InspectorOptionsColumn
                            // eslint-disable-next-line react/no-array-index-key
                            key={index}
                            options={options}
                            contextGroup={contextGroup}
                            selectedItem={selectedItem as Array<number>}
                            trailingSelection={trailingSelection}
                            filteredItems={filteredItems as Array<Array<number>>}
                            onSelect={onSelect}
                            renderItemContent={renderItemContent}
                        />
                    );
                })}
                {appendEmptyColumn
                    && <div className="jsonschema-inspector-column-placeholder" />}
            </div>
        );
    }

    static propTypes = {
        columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(true))).isRequired,
        appendEmptyColumn: PropTypes.bool,
        renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
    };
    
    static defaultProps: ColViewDefaultProps = {
        appendEmptyColumn: false,
        renderItemContent: null
    };
}

export default InspectorColView;
