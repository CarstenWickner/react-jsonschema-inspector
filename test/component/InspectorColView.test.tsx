import * as React from "react";
import { mount, shallow } from "enzyme";

import { ColViewProps, InspectorColView } from "../../src/component/InspectorColView";
import { JsonSchemaGroup } from "../../src/model/JsonSchemaGroup";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const component = shallow(
            <InspectorColView
                columnData={[
                    {
                        items: {
                            "Item One": new JsonSchemaGroup(),
                            "Item Two": new JsonSchemaGroup()
                        },
                        selectedItem: "Item One",
                        onSelect: (): void => {}
                    },
                    {
                        options: {
                            groupTitle: "one of",
                            options: [{}, {}]
                        },
                        contextGroup: new JsonSchemaGroup(),
                        selectedItem: [0],
                        trailingSelection: true,
                        onSelect: (): void => {}
                    },
                    {
                        items: {
                            "Item One-Two-One": new JsonSchemaGroup(),
                            "Item One-Two-Two": new JsonSchemaGroup()
                        },
                        onSelect: (): void => {}
                    }
                ]}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("with empty column", () => {
        const component = shallow(
            <InspectorColView
                columnData={[
                    {
                        items: {
                            "Item One": new JsonSchemaGroup(),
                            "Item Two": new JsonSchemaGroup()
                        },
                        selectedItem: "Item Two",
                        trailingSelection: true,
                        onSelect: (): void => {}
                    }
                ]}
                appendEmptyColumn
            />
        );
        expect(component.exists(".jsonschema-inspector-column-placeholder")).toBe(true);
    });
});
describe("update according to prop changes", () => {
    const singleColumnData: ColViewProps["columnData"] = [
        {
            items: {
                "Item One": new JsonSchemaGroup(),
                "Item Two": new JsonSchemaGroup()
            },
            onSelect: (): void => {}
        }
    ];
    const doubleColumnData: ColViewProps["columnData"] = [
        {
            items: {
                "Item One": new JsonSchemaGroup(),
                "Item Two": new JsonSchemaGroup()
            },
            selectedItem: "Item Two",
            trailingSelection: true,
            onSelect: (): void => {}
        },
        {
            items: {
                "Item Two-One": new JsonSchemaGroup(),
                "Item Two-Two": new JsonSchemaGroup()
            },
            onSelect: (): void => {}
        }
    ];

    it("setting root selection (while no empty column was displayed)", () => {
        const component = mount(<InspectorColView columnData={singleColumnData} />);
        // simulate selection in root column
        component.setProps({
            columnData: doubleColumnData
        });
        expect(component.find(doubleColumnData[0]).exists()).toBe(true);
        expect(component.find(doubleColumnData[1]).exists()).toBe(true);
    });
    it("setting root selection (while empty column was displayed)", () => {
        const component = mount(<InspectorColView columnData={singleColumnData} appendEmptyColumn />);
        // simulate selection in root column
        component.setProps({
            columnData: doubleColumnData
        });
        expect(component.find(doubleColumnData[0]).exists()).toBe(true);
        expect(component.find(doubleColumnData[1]).exists()).toBe(true);
    });
    it("clearing root selection", () => {
        const component = mount(<InspectorColView columnData={doubleColumnData} />);
        // simulate clearing the selection again (thereby adding an empty column)
        component.setProps({
            columnData: singleColumnData,
            appendEmptyColumn: true
        });
        const rootColumn = component.find(singleColumnData[0]);
        expect(rootColumn.exists()).toBe(true);
        expect(Array.isArray(rootColumn)).toBe(false);
    });
});
