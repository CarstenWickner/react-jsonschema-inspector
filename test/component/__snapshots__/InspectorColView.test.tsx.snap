// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-colview"
  tabIndex={-1}
>
  <InspectorColumn
    items={
      Object {
        "Item One": JsonSchemaGroup {
          "entries": Array [],
        },
        "Item Two": JsonSchemaGroup {
          "entries": Array [],
        },
      }
    }
    key="0"
    onSelect={[Function]}
    selectedItem="Item One"
  />
  <InspectorOptionsColumn
    contextGroup={
      JsonSchemaGroup {
        "entries": Array [],
      }
    }
    key="1"
    onSelect={[Function]}
    options={
      Object {
        "groupTitle": "one of",
        "options": Array [
          Object {},
          Object {},
        ],
      }
    }
    selectedItem={
      Array [
        0,
      ]
    }
    trailingSelection={true}
  />
  <InspectorColumn
    items={
      Object {
        "Item One-Two-One": JsonSchemaGroup {
          "entries": Array [],
        },
        "Item One-Two-Two": JsonSchemaGroup {
          "entries": Array [],
        },
      }
    }
    key="2"
    onSelect={[Function]}
  />
</div>
`;
